llvm/lib/Target/AIE/AIEPostPipeliner.h

```diff
@@ -43,14 +45,77 @@ class NodeInfo {
   int ModuloCycle = 0;
   // Cycle / II
   int Stage = 0;
+
   // The earliest cycle at which this can be scheduled to meet latencies
   // This includes the lowerbound of the modulo condition, i.e.
   // Earliest(N) >= Cycle(N - NInstr) + II
   int Earliest = 0;
-  // For an LCD K1 -> K2, this holds II + Earliest(K2 - NInstr) - Latency(LCD)
-  // Instructions with lower Latest have higher priority in the
-  // top down scheduling
-  int Latest = 0;
+
+  // The latest cycle at which this can be scheduled. This is a negative value
+  // relative to the length of the linear schedule.
+  // So -1 is the last cycle of the linear schedule, -Length is the first cycle
+  // of the linear schedule. Note that this length is usually rounded up to
+  // the next multiple of the initiation interval
+  int Latest = -1;
+
+  // These are the values of Earliest and Latest as computed from the a-priori
+  // computations. During scheduling Earliest and Latest may be adjusted to
+  // more accurate values. The two values are cached here to facilitate cheaper
+  // reset before trying a new strategy for the same II.
+  int StaticEarliest = 0;
+  int StaticLatest = -1;
+
+  // Slots necessary for this instruction.
+  SlotCounts Slots;
+
+  // Record critical path components
+  // The Pred/Succ that pushed my Earliest/Latest
+  std::optional<int> LastEarliestPusher;
+  std::optional<int> LastLatestPusher;
+  // The number of Succs/Preds whose Earliest/Latest I have pushed.
+  int NumPushedEarliest = 0;
+  int NumPushedLatest = 0;
+
+  // Latest corrected by taking Earliest of an LCD successor into account
+  int LCDLatest = -1;
```

Comment:
Any reason not to have a symmetric `LCDEarliest` for bottom-up scheduling?

Reply:
But I guess we would not have a strategy to make use of it anyway?

Reply:
Correct. Leaving it for now, but we should keep in mind that LcdLatest is asymmetric wrt TopDown now.